FreeRTOS/FreeRTOS-Kernel
# Add SMP template port and example

Add SMP template port and example

Description
-----------
The SMP template port shares of the same template port for single core. The port macros and implementation differences are separated by compile option `configNUMBER_OF_CORES`. Sharing the same port template helps user to identify the port differences between single core and SMP. A sample configuration file for SMP with minimal configuration change is also provided. The example project with cmake is updated to consume the template port and sample configuration to provide compilable project.

The configuration added for SMP is listed below
```c
/******************************************************************************/
/* Scheduling behaviour related definitions. **********************************/
/******************************************************************************/

/* Set configNUMBER_OF_CORES to greater than 1 to enable running one instance of
 * FreeRTOS kernel to schedule tasks across multiple identical processor cores. */
#define configNUMBER_OF_CORES    2

/******************************************************************************/
/* Hook and callback function related definitions. ****************************/
/******************************************************************************/

/* Set the following configUSE_* constants to 1 to include the named hook
 * functionality in the build.  Set to 0 to exclude the hook functionality from the
 * build.  The application writer is responsible for providing the hook function
 * for any set to 1.  See https://www.freertos.org/a00016.html */
#define configUSE_PASSIVE_IDLE_HOOK    0
```


As single core template port, the port is expected to be used for:
* Provide a starting point for a SMP port
* Provide a compilable project for static analysis tool. Since the sample configuration only enable a small set of features to reduce project complexity, a project for static analysis tool may use another configuration to enable more features for coverage. 




Test Steps
-----------
To build the example project for SMP template port:
```
cd examples\cmake_example
cmake -S . -B build -DFREERTOS_SMP_EXAMPLE=1
cd build
make
```
`smp_example` should be generated in the build folder.

Checklist:
----------



Related Issue
-----------



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
